projects/solar/src/solar_fsm.c

```diff
@@ -0,0 +1,92 @@
+#include "solar_fsm.h"
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "can.h"
+#include "can_msg_defs.h"
+#include "can_unpack.h"
+#include "drv120_relay.h"
+#include "event_queue.h"
+#include "exported_enums.h"
+#include "fsm.h"
+#include "solar_events.h"
+#include "status.h"
+
+FSM_DECLARE_STATE(state_relay_open);
+FSM_DECLARE_STATE(state_relay_closed);
+
+static bool prv_relay_open_fault_guard(const Fsm *fsm, const Event *e, void *context) {
+  SolarFsmStorage *storage = context;
+  EESolarFault fault = GET_FAULT_FROM_EVENT(*e);
+  for (uint8_t i = 0; i < storage->num_relay_open_faults; i++) {
+    if (storage->relay_open_faults[i] == fault) {
+      return true;
+    }
+  }
+  return false;
+}
+
+FSM_STATE_TRANSITION(state_relay_open) {
+  FSM_ADD_TRANSITION(SOLAR_COMMAND_EVENT_CLOSE_RELAY, state_relay_closed);
+}
+
+FSM_STATE_TRANSITION(state_relay_closed) {
+  FSM_ADD_TRANSITION(SOLAR_COMMAND_EVENT_OPEN_RELAY, state_relay_open);
+  FSM_ADD_GUARDED_TRANSITION(SOLAR_FAULT_EVENT, prv_relay_open_fault_guard, state_relay_open);
+}
+
+static void prv_open_relay(Fsm *fsm, const Event *e, void *context) {
+  drv120_relay_open();
+}
+
+static void prv_close_relay(Fsm *fsm, const Event *e, void *context) {
+  drv120_relay_close();
+}
+
+static StatusCode prv_handle_relay_state_can(const CanMessage *msg, void *context,
+                                             CanAckStatus *ack) {
+  // mirror the battery relay state
+  uint16_t relay_mask, relay_state;
+  CAN_UNPACK_SET_RELAY_STATES(msg, &relay_mask, &relay_state);
+
+  if (relay_mask & (1 << EE_RELAY_ID_BATTERY)) {
+    if (relay_state & (EE_RELAY_STATE_CLOSE << EE_RELAY_ID_BATTERY)) {
```

Comment:
why not just 1 like the previous line ... if you want to check if the value matches EE_RELAY_STATE_CLOSE then you should do something like:
```c
(relay_state & (1 << EE_RELAY_ID_BATTERY)) >> EE_RELAY_ID_BATTERY  ==  EE_RELAY_STATE_CLOSE
```
(since changing to EE_RELAY_STATE_OPEN breaks the check)
Also technically [code style](https://uwmidsun.atlassian.net/wiki/spaces/ELEC/pages/9306218/Coding+Standards) for both require the `!= 0`. See "Comparison with NULL / 0" heading